src/libaktualizr/primary/sotauptaneclient.cc

```diff
@@ -837,18 +849,27 @@ result::Install SotaUptaneClient::uptaneInstall(const std::vector<Uptane::Target
     return result;
   }
 
+  const Uptane::EcuSerial &primary_ecu_serial = primaryEcuSerial();
+
   // Recheck the downloaded update hashes.
   for (const auto &update : updates) {
-    if (package_manager_->verifyTarget(update) != TargetStatus::kGood) {
-      result.dev_report = {false, data::ResultCode::Numeric::kInternalError, ""};
-      storage->storeDeviceInstallationResult(result.dev_report, "Downloaded target is invalid", correlation_id);
-      sendEvent<event::AllInstallsComplete>(result);
-      return result;
+    if (update.IsForEcu(primary_ecu_serial) || !update.IsOstree()) {
+      // download binary images for any target, for both Primary and Secondary
+      // download an ostree revision just for Primary, Secondary will do it by itself
+      // Primary cannot verify downloaded OSTree targets for Secondaries,
+      // Downloading of Secondary's ostree repo revision to the Primary's can fail
+      // if they differ signficantly as ostree has a certain cap/limit of the diff it pulls
```

Comment:
Can you explain this bit about "ostree has a certain cap/limit of the diff it pulls"? I wasn't aware of that.

To me, the comment about "Downloading of Secondary's ostree repo revision to the Primary" is a bit confusing, because the larger conceptual problem is that the Primary shouldn't be downloading objects for the Secondary if it has no use for them and they would just waste space.

Reply:
>  the Primary shouldn't be downloading objects for the Secondary if it has no use for them and they would just waste space.

That's exactly what we need this if statement `if (update.IsForEcu(primary_ecu_serial) || !update.IsOstree())` for, i.e. to avoid downloading of an ostree revision aimed for a secondary's repo to a primary's ostree repo. The purpose of this comment is to justify this of statement.

Reply:
Fair enough, I was just a bit thrown off. But is it really true that ostree limits the size of a diff it pulls?

Reply:
> Can you explain this bit about "ostree has a certain cap/limit of the diff it pulls"? I wasn't aware of that.

`Dec 12 17:27:56 qemux86-64 aktualizr while pulling image: 0 Writing content object: min-free-space-percent '3%' would be exceeded, at least 12.2 MB requested

Error message while pulling secondary's ostree repo version/revision/commit-objects to the primary's one if there is big diff between two of them`

https://github.com/ostreedev/ostree/blob/1645416aac802d9dde204117d8fd6352d3958f4a/src/libostree/ostree-repo-commit.c#L442

Reply:
Gotcha. That error is just because the primary is running out of space, though; that isn't specifically dependent on the size of the diff. I've seen something like "error 34 need more input" which I was hoping was related, but I still don't always *really* understand why I see that.

Reply:
> That error is just because the primary is running out of space, though; that isn't specifically dependent on the size of the diff.

My understanding is that the error depends on both an available free space on a storage and a diff size (https://salsa.debian.org/debian/ostree/commit/1f5ce1a9f789d9c0de5d6fbdf79540bf71c5bc9b,  "specifies a minimum
        percentage of total space (in blocks) in the underlying filesystem to
        keep free. The default value is 3
")

Reply:
also, I get an error when I am trying to pull an ostree revision of the qemu secondary/primary to a local ostree generated by the makephysical tool.